spring-cloud-function-context/src/main/java/org/springframework/cloud/function/cloudevent/CloudEventMessageUtils.java

```diff
@@ -226,15 +221,16 @@ public static Map<String, Object> getAttributes(Message<?> message) {
 
 	/**
 	 * This method does several things.
-	 * First in canonicalizes Cloud Events attributes ensuring that they all prefixed
+	 * First it canonicalizes Cloud Events attributes ensuring that they are all prefixed
 	 * with 'ce-' prefix regardless where they came from.
-	 * It also transforms structured-mode Cloud Event to binary-mode and then it canonicalizes attributes
+	 * It also transforms structured-mode Cloud Event to binary-mode and then canonicalizes attributes
 	 * as well as described in the previous sentence.
 	 */
 	@SuppressWarnings("unchecked")
 	static Message<?> toCanonical(Message<?> inputMessage, MessageConverter messageConverter) {
-		Map<String, Object> headers = (Map<String, Object>) ReflectionUtils.getField(MESSAGE_HEADERS, inputMessage.getHeaders());
-		canonicalizeHeaders(headers, false);
+		inputMessage = canonicalizeHeadersWithPossibleCopy(inputMessage);
```

Comment:
This is in effect what we were doing w/ reflection, albeit previously directly on the immutable message. Also, an optimization made is we only copy when something is actually changed.